### PR TITLE
Rework Code Quality repost generation: fingerprint no longer depends on line number

### PR DESCRIPTION
**Pull request by JokeWaumans**
Add a class `Finding` to store the data of a finding of the Code Quality report.
This makes it easier to generate a unique fingerprint for each finding.

### CODE REVIEW

File: src/mlx/warnings/code_quality.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+from pathlib import Path
+
+
+class Finding:
+    """Code quality violation"""
+
+    """A dictionary mapping fingerprints (unique hashes) to instances of Finding"""
+    fingerprints = {}
+
+    _severity = ""
```
Comment (JasperCraeghs) [inline src/mlx/warnings/code_quality.py:12]:
```suggestion
    _severity = "info"
```
You'll no longer need an `else`-statement to set a fallback value

File: src/mlx/warnings/code_quality.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    _path = ""
+    _description = ""
```
Comment (JasperCraeghs) [inline src/mlx/warnings/code_quality.py:14]: I think that description doesn't need a property and setter, so no need for a protected counterpart then.

File: src/mlx/warnings/code_quality.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        2. The base string is hashed to produce an initial hash.
+        3. If the hash already exists (hash collision), a new hash is generated by concatenating the base string
+           with the current hash and hashing the result.
+        4. Step 3 is repeated until a unique hash is obtained.
+        """
+        hashable_string = f"{self.severity}{self.path}{self.description}"
+        new_hash = hashlib.md5(str(hashable_string).encode('utf-8')).hexdigest()
+        while new_hash in self.fingerprints and self.fingerprints[new_hash] != self:
+            new_hash = hashlib.md5(f"{hashable_string}{new_hash}".encode('utf-8')).hexdigest()
+        self.fingerprints[new_hash] = self
```
Comment (JasperCraeghs) [inline src/mlx/warnings/code_quality.py:38]: Can you explain how this works? Don't we need `type(self).fingerprints` instead? https://realpython.com/python-classes/#class-attributes

File: src/mlx/warnings/warnings.py [hunk 4]
```diff
@@ -277,6 +278,7 @@ def warnings_wrapper(args):
                         help='Possible not-used flags from above are considered as command flags')
 
     args = parser.parse_args(args)
+    Finding.fingerprints = {}  # Make sure the class variable is an empty dict
```
Comment (JasperCraeghs) [inline src/mlx/warnings/warnings.py:281]: Why here? Did you run into issues when executing multiple test cases?

File: tests/test_coverity.py [hunk 5]
```diff
@@ -66,7 +66,7 @@ def test_code_quality_without_config(self):
         self.assertTrue(filecmp.cmp(out_file, ref_file))
 
     def test_code_quality_with_config(self):
-        filename = 'coverity_cq.json'
+        filename = 'coverity_cq_with_config.json'
```
Comment (JasperCraeghs) [inline tests/test_coverity.py:69]: The new file is identical, isn't it?

**Review by JasperCraeghs — CHANGES_REQUESTED** [review]

File: src/mlx/warnings/code_quality.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+    """A dictionary mapping fingerprints (unique hashes) to instances of Finding"""
+    fingerprints = {}
+
+    _severity = ""
+    _path = ""
+    _description = ""
+    _line = 1
+    _column = 1
+    _hashable_string = ""
```
[the remark this answers is not in the document]
Comment (JokeWaumans) [reply src/mlx/warnings/code_quality.py]: Resovled in d071251c6f0084f8ee4dd44312d847f61f755cf7

File: src/mlx/warnings/warnings.py (hunk 4 above)
Comment (JokeWaumans) [reply src/mlx/warnings/warnings.py]: Resolved in e5e42e743633ba586bab1c0fad13c53371ff7f67

File: src/mlx/warnings/code_quality.py (hunk 2 above)
Comment (JokeWaumans) [reply src/mlx/warnings/code_quality.py]: Resolved in 2043fc123c1c23906a2242ad83aa55aed651ee3d

**Review by JasperCraeghs — APPROVED** [review]

**Pull request merged**